### PR TITLE
Arduino STM32 Core support

### PR DESCRIPTION
**Pull request by ubis**
In order to make this work, 2 packages needs to be prepared and downloaded: [stm32 core](https://github.com/stm32duino/Arduino_Core_STM32) and [CMSIS 5](https://github.com/ARM-software/CMSIS_5)

### CODE REVIEW

File: builder/frameworks/arduino/stm32core.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+library = "arm_" + board.get("build.cpu")[:6] + m
+
+# mcu's that require additional flags
+mcu_list = [
+    "f429", "l496", "f302", "f303re",
+    "f401", "f411", "f446", "l476",
+    "l432", "f407", "f746", "l475"
+]
+
+for item in mcu_list:
```
Comment (ivankravets) [inline builder/frameworks/arduino/stm32core.py]: You can replace it with 
```
if any(item in mcu_type for item mcu_list):
    env.Append(..
```

File: builder/frameworks/arduino/stm32duino.py [hunk 2]
```diff
@@ -26,7 +26,35 @@
...
+    "l432kc", "l476rg", "l496zg", "l475vg"
+]
+
+mcu_type = env.BoardConfig().get("build.mcu")
+official = False
+
+if not "f103" in mcu_type and not "f407" in mcu_type:
+    official = True
+
+if "STM32_OFFICIAL_CORE" in env['CPPDEFINES']:
```
Comment (ivankravets) [inline builder/frameworks/arduino/stm32duino.py]: Could you replace this macro with `PIO_FRAMEWORK_ARDUINO_OFFICIAL` to make it clear that it is related just to PlatformIO?

**Comment by ivankravets** [comment]
Could you also extend Arduino examples with a few official boards?

- https://github.com/platformio/platform-ststm32/blob/develop/examples/arduino-blink/platformio.ini
- https://github.com/platformio/platform-ststm32/blob/develop/examples/arduino-external-libs/platformio.ini
- https://github.com/platformio/platform-ststm32/blob/develop/examples/arduino-internal-libs/platformio.ini

**Comment by ivankravets** [comment]
> In order to make this work, 2 packages needs to be prepared and downloaded: stm32 core and CMSIS 5

Could you prepeare a test package and share somewhere? I jsut want to take a look at the file structure.

**Comment by ubis** [comment]
Sorry for the late reply, I was quite busy on the weekend 🙂
This is the test [package](https://ufile.io/4v13d). Blink example was successful, `external libs` was not, until I renamed `*.cpp` into `*.ino`. Seems like there's some additional stuff with `*.ino`. And `internal libs` did not compile with stm32 core, because they do not have `RTClock` library in their core yet, so I have not committed changes for these examples.

**Comment by ivankravets** [comment]
Thank very much for your work! I've just done some fixes and all tests are passed:
- https://travis-ci.org/platformio/platform-ststm32/builds/443789142

1. I renamed build scripts. Please pull all changes
2. New package has been deployed to our CDN
3. I used the latest CMSIS 5.3.0 from GitHub and removed ARM compiler related libraries and files. The final size of package was reduced from 487Mb to 115Mb
4. I switched to v1.3.0 official release of https://github.com/stm32duino/Arduino_Core_STM32/releases/tag/1.3.0. We don't support BETA software.
5. I see a lot of boards were not added to our registry. See https://github.com/stm32duino/Arduino_Core_STM32/blob/1.3.0/boards.txt

Could you add support for all these boards? If manifest does not exist, please create it following the same rules which we used to create previous manifests.

Thanks in advance!

**Comment by ivankravets** [comment]
Ah... The last! I really don't like an idea with using `PIO_FRAMEWORK_ARDUINO_OFFICIAL`. I see that Roger's repo is not maintained well. I propose to switch to official STM32Duino and use it by default for all boards. The only if board is not yet supported by official core, then use Roger's core. 

What do you think?

**Comment by ubis** [comment]
> I see a lot of boards were not added to our registry

Hmm, what do you mean? All boards were added 😉

As for `PIO_FRAMEWORK_ARDUINO_OFFICIAL` it is intented only for `F103` and `F407` series, because both frameworks support these boards and I did not want to not make any incompatibilities when suddenly framework would be switched and users could get some errors or so.

> I propose to switch to official STM32Duino and use it by default for all boards.

At the moment it is default for all STM32 official core supported boards, except `F103/F407`. There are more `F103` boards supported in Roger's core than STM32's official.

**Comment by ivankravets** [comment]
> Hmm, what do you mean? All boards were added 

* Nucleo L496ZG-P
* Nucleo L031K6
* DISCO_F407VG

>  I did not want to not make any incompatibilities

Which incompatibilities could we have? This Arduino API, should be the same in the both cores. In any case, we will soon deprecate Roger's core and leave only official core.

I propose to add `PIO_FRAMEWORK_ARDUINO_LEGACY_CORE` macro and use it only in case when developer asks or this boars is only supported in legacy core.

What do you think?

**Comment by ubis** [comment]
> Nucleo L496ZG-P

It is the same as L496ZG, same flags, same definitions.

> Nucleo L031K6

Hmm.. It is supported already.

> DISCO_F407VG

It is supported by using `PIO_FRAMEWORK_ARDUINO_OFFICIAL` as I have said earlier. It is the same for `F103/F407` boards.

> Which incompatibilities could we have?

Well, when I have tried examples, it seems that there are more libraries inside Roger's core.

> I propose to add `PIO_FRAMEWORK_ARDUINO_LEGACY_CORE`

Sounds good to me.

**Comment by ivankravets** [comment]
@ubis could you help us with these 2 things? We would like to make a final release. Thanks!

**Pull request merged**